src/poke_env/environment/double_battle.py

```diff
@@ -0,0 +1,425 @@
+                                        "move such as copycat or me first. If that is "
+                                        "not the case, please make sure there is an "
+                                        "explanation for this behavior or report it "
+                                        "if it is an error.",
+                                        move["id"],
+                                        self.battle_tag,
+                                        active_pokemon.species,
+                                    )
+                                move = Move(move["id"])
+                                self._available_moves[active_pokemon_number].append(
+                                    move
+                                )
+                            except AttributeError:
+                                pass
+
+                if active_request.get("canMegaEvo", False):
+                    self._can_mega_evolve[active_pokemon_number] = True
+                if active_request.get("canZMove", False):
+                    self._can_z_move[active_pokemon_number] = True
+                if active_request.get("canDynamax", False):
+                    self._can_dynamax[active_pokemon_number] = True
+                if active_request.get("maybeTrapped", False):
+                    self._maybe_trapped[active_pokemon_number] = True
+
+        for trapped, pokemon_index in enumerate(self.trapped):
+            if not trapped:
+                for pokemon in side["pokemon"]:
+                    if pokemon:
+                        pokemon = self._team[pokemon["ident"]]
+                        if not pokemon.active and not pokemon.fainted:
+                            self._available_switches[pokemon_index].append(pokemon)
+
+    def _switch(self, pokemon, details, hp_status):
+        pokemon_identifier = pokemon.split(":")[0][:3]
+        player_identifier = pokemon_identifier[:2]
+        team = (
+            self._active_pokemon
+            if player_identifier == self._player_role
+            else self._opponent_active_pokemon
+        )
+        pokemon_out = team.pop(pokemon_identifier, None)
+        if pokemon_out is not None:
+            pokemon_out._switch_out()
+        pokemon_in = self.get_pokemon(pokemon, details=details)
+        pokemon_in._switch_in()
+        pokemon_in._set_hp_status(hp_status)
+        team[pokemon_identifier] = pokemon_in
+
+    def _swap(self, pokemon, slot):
+        player_identifier = pokemon.split(":")[0][:2]
+        team = (
+            self._active_pokemon
+            if player_identifier == self.player_role
+            else self._opponent_active_pokemon
+        )
+        slot_a = f"{player_identifier}a"
+        slot_b = f"{player_identifier}b"
+        team[slot_a], team[slot_b] = team[slot_b], team[slot_a]
+
+    def get_possible_showdown_targets(self, move: Move, dynamax=False) -> List[int]:
```

Comment:
This should be in `Move` and broken up into several smaller methods

Reply:
this requires access to `DoubleBattle` object state, so I can't see this getting moved somewhere else, unless you want to pass the battle object as an argument i guess